### PR TITLE
fix(extensions): define platform info to prevent renderer crash

### PR DESCRIPTION
**Pull request by samuelmaddock**
#### Description of Change

Fully implements `chrome.runtime.getPlatformInfo` such that the calling renderer won't crash/hang.

The implementation is copied from [chrome/browser/extensions/api/runtime/chrome_runtime_api_delegate.cc](https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/extensions/api/runtime/chrome_runtime_api_delegate.cc;l=285;drc=89ce2d1e6202423c9c0f0c83cec1f198cf55eded).

Ref #19447

cc @nornagon @sentialx 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed `chrome.runtime.getPlatformInfo` crashing the background process upon being invoked.

### CODE REVIEW

File: spec-main/fixtures/extensions/chrome-runtime/main.js [hunk 1]
```diff
@@ -1,6 +1,40 @@
 /* eslint-disable */
-document.documentElement.textContent = JSON.stringify({
-  manifest: chrome.runtime.getManifest(),
-  id: chrome.runtime.id,
-  url: chrome.runtime.getURL('main.js')
+
```
Comment (samuelmaddock) [inline spec-main/fixtures/extensions/chrome-runtime/main.js:2]: I converted this script to another pattern we use such that each API can be tested in isolation.

**Comment by sentialx** [comment]
LGTM

**Comment by nornagon** [comment]
Looks like you're missing a dependency declaration in the BUILD.gn file:

```
ERROR at //electron/shell/browser/extensions/api/runtime/electron_runtime_api_delegate.cc:10:11: Can't include this header from here.
#include "components/update_client/update_query_params.h"
          ^---------------------------------------------
The target:
  //electron:electron_lib
is including a file from the target:
  //components/update_client:update_client
```

While it's a little questionable for us to be making a dependency on something called "update client" (given we don't use any of the actual functionality of that component), we're already transitively dependent on it via `//extensions/browser`, so adding it as an explicit dependency seems OK.

**Comment by nornagon** [comment]
Also, I'm surprised this crashed before. Can you provide some more info about the crash this fixes?

**Comment by samuelmaddock** [comment]
> 
> 
> Also, I'm surprised this crashed before. Can you provide some more info about the crash this fixes?

@nornagon what seems to happen is the API attempts to use `PlatformInfo` in a response, but doesn't handle it well when some of the fields aren't initialized. Specifically the call to [runtime::GetPlatformInfo::Results::Create(info)](https://source.chromium.org/chromium/chromium/src/+/master:extensions/browser/api/runtime/runtime_api.cc;l=723;drc=3003f448ef9a2d6407d1369d2aa8543f7a0a7c63). Then `runtime::ToString` is called on uninitialized data in a [PlatformInfo generated function.](https://source.chromium.org/chromium/chromium/src/+/master:out/Debug/gen/extensions/common/api/runtime.cc;l=225;drc=3003f448ef9a2d6407d1369d2aa8543f7a0a7c63)

File: shell/browser/extensions/api/runtime/electron_runtime_api_delegate.cc [hunk 2]
```diff
@@ -42,10 +43,57 @@ bool ElectronRuntimeAPIDelegate::CheckForUpdates(
...
+  } else if (strcmp(nacl_arch, "x86-64") == 0) {
+    info->nacl_arch = extensions::api::runtime::PLATFORM_NACL_ARCH_X86_64;
+  } else if (strcmp(nacl_arch, "mips32") == 0) {
+    info->nacl_arch = extensions::api::runtime::PLATFORM_NACL_ARCH_MIPS;
+  } else if (strcmp(nacl_arch, "mips64") == 0) {
+    info->nacl_arch = extensions::api::runtime::PLATFORM_NACL_ARCH_MIPS64;
+  } else {
+    NOTREACHED();
+    return false;
+  }
```
Comment (nornagon) [inline shell/browser/extensions/api/runtime/electron_runtime_api_delegate.cc:95]: We never support NaCl. Can we change this to:
```suggestion
  info->nacl_arch = extensions::api::runtime::PLATFORM_NACL_ARCH_NONE;
```

File: spec-main/extensions-spec.ts [hunk 3]
```diff
@@ -184,28 +184,38 @@ describe('chrome extensions', () => {
...
-        w.loadURL(url);
-        await emittedOnce(w.webContents, 'dom-ready');
-        content = JSON.parse(await w.webContents.executeJavaScript('document.documentElement.textContent'));
-        expect(content).to.be.an('object');
-      } finally {
-        w.destroy();
-      }
+      w = new BrowserWindow({ show: false, webPreferences: { session: customSession, nodeIntegration: true } });
+      w.loadURL(url);
+      await emittedOnce(w.webContents, 'dom-ready');
```
Comment (nornagon) [inline spec-main/extensions-spec.ts:199]: do we need to wait for `dom-ready` here or can we just `await w.loadURL`?

File: spec-main/fixtures/extensions/chrome-runtime/main.js (hunk 1 above)
Comment (nornagon) [reply spec-main/fixtures/extensions/chrome-runtime/main.js:2]: Nice change 👍

File: shell/browser/extensions/api/runtime/electron_runtime_api_delegate.cc (hunk 2 above)
Comment (samuelmaddock) [reply shell/browser/extensions/api/runtime/electron_runtime_api_delegate.cc]: So the extensions code doesn't seem to like using this value. By doing so, the renderer crashes and devtools gets disconnected. 😞

I reverted the suggested changes, but removed unsupported platforms (mips, mips64).

File: spec-main/extensions-spec.ts (hunk 3 above)
Comment (samuelmaddock) [reply spec-main/extensions-spec.ts]: Nope. I've updated extension tests to only await `dom-ready` when necessary.

**Review by gittysachin — COMMENTED** [review]
So far, LGTM

File: shell/browser/extensions/api/runtime/electron_runtime_api_delegate.cc (hunk 2 above)
Comment (nornagon) [reply shell/browser/extensions/api/runtime/electron_runtime_api_delegate.cc]: hm, ok. /shrug then

**Review by nornagon — APPROVED** [review]

**Pull request merged**